apps/arweave/src/ar_mining_server.erl

```diff
@@ -228,89 +174,157 @@ handle_cast({post_solution, Solution}, State) ->
+	?LOG_DEBUG([{event, mining_debug_skipping_vdf_output}]),
 	{noreply, State};
-handle_info({event, nonce_limiter, {computed_output, _}},
-		#state{ session = #mining_session{ paused = true } } = State) ->
-	{noreply, State};
-handle_info({event, nonce_limiter, {computed_output, Args}},
-		#state{ task_queue = Q } = State) ->
+handle_info({event, nonce_limiter, {computed_output, Args}}, State) ->
 	{SessionKey, StepNumber, Output, PartitionUpperBound} = Args,
+
 	true = is_integer(StepNumber),
 	ar_mining_stats:vdf_computed(),
-	Task = {computed_output, {SessionKey, StepNumber, Output, PartitionUpperBound}},
-	Q2 = gb_sets:insert({priority(nonce_limiter_computed_output, StepNumber), make_ref(),
-			Task}, Q),
-	prometheus_gauge:inc(mining_server_task_queue_len, [computed_output]),
-	{noreply, State#state{ task_queue = Q2 }};
-handle_info({event, nonce_limiter, _}, State) ->
+
+	State2 = update_upper_bound(PartitionUpperBound, State),
+
+	{NextSeed, StartIntervalNumber, NextVDFDifficulty} = SessionKey,
+	{Workers, State3} = 
+		case get_workers(SessionKey, State2) of
+			[] ->
+				RefreshedState = refresh_workers(State2),
+				{get_workers(SessionKey, RefreshedState), RefreshedState};
+			Workers2 ->
+				{Workers2, State2}
+		end,
+	
+	case Workers of
+		[] ->
+			?LOG_DEBUG([{event, mining_debug_skipping_vdf_output},
+				{step_number, StepNumber},
+				{session_key, ar_nonce_limiter:encode_session_key(SessionKey)}]),
+			ok;
+		_ ->
+			Candidate = #mining_candidate{
+				session_key = SessionKey,
+				next_seed = NextSeed,
+				next_vdf_difficulty = NextVDFDifficulty,
+				start_interval_number = StartIntervalNumber,
+				step_number = StepNumber,
+				nonce_limiter_output = Output,
+				partition_upper_bound = PartitionUpperBound
+			},
+			N = distribute_output(Workers, Candidate),
+			?LOG_DEBUG([{event, mining_debug_processing_vdf_output}, {found_io_threads, N},
+				{step_number, StepNumber}, {output, ar_util:safe_encode(Output)},
+				{start_interval_number, StartIntervalNumber},
+				{session_key, ar_nonce_limiter:encode_session_key(SessionKey)}])
+	end,
+	{noreply, State3};
+
+handle_info({event, nonce_limiter, Message}, State) ->
+	?LOG_DEBUG([{event, mining_debug_skipping_nonce_limiter}, {message, Message}]),
 	{noreply, State};
 
 handle_info(Message, State) ->
-	?LOG_WARNING("event: unhandled_info, message: ~p", [Message]),
+	?LOG_WARNING([{event, unhandled_info}, {module, ?MODULE}, {message, Message}]),
 	{noreply, State}.
 
-terminate(_Reason, #state{ hashing_threads = HashingThreads }) ->
-	[Thread ! stop || Thread <- queue:to_list(HashingThreads)],
+terminate(_Reason, _State) ->
 	ok.
 
 %%%===================================================================
 %%% Private functions.
 %%%===================================================================
+update_upper_bound(PartitionUpperBound, State) ->
+	#state{ partition_upper_bound = CurrentUpperBound } = State,
```

Comment:
Do we still need partition_upper_bound in State?

Also, what do you think about naming the function something like `update_largest_seen_upper_bound` to make it clear it's not the bound used for picking recall ranges?

Reply:
Good point - we don't! removed. https://github.com/ArweaveTeam/arweave/pull/503/commits/e34a6961ddc613b2febc2eaaebba3e88c568ac41